pkg/operators/wasm/datasource.go

```diff
@@ -0,0 +1,525 @@
+			wapi.ValueTypeI32, // Priority
+			wapi.ValueTypeI64, // CallbackID
+		},
+		[]wapi.ValueType{wapi.ValueTypeI32}, // Error
+	)
+
+	exportFunction(env, "dataSourceGetField", i.dataSourceGetField,
+		[]wapi.ValueType{
+			wapi.ValueTypeI32, // DataSource
+			wapi.ValueTypeI64, // FieldName
+		},
+		[]wapi.ValueType{wapi.ValueTypeI32}, // Accessor
+	)
+
+	exportFunction(env, "dataSourceAddField", i.dataSourceAddField,
+		[]wapi.ValueType{
+			wapi.ValueTypeI32, // DataSource
+			wapi.ValueTypeI64, // FieldName
+			wapi.ValueTypeI32, // FieldKind
+		},
+		[]wapi.ValueType{wapi.ValueTypeI32}, // Accessor
+	)
+
+	exportFunction(env, "dataSourceNewPacketSingle", i.dataSourceNewPacketSingle,
+		[]wapi.ValueType{wapi.ValueTypeI32}, // DataSource
+		[]wapi.ValueType{wapi.ValueTypeI32}, // Packet
+	)
+
+	exportFunction(env, "dataSourceNewPacketArray", i.dataSourceNewPacketArray,
+		[]wapi.ValueType{wapi.ValueTypeI32}, // DataSource
+		[]wapi.ValueType{wapi.ValueTypeI32}, // Packet
+	)
+
+	exportFunction(env, "dataSourceEmitAndRelease", i.dataSourceEmitAndRelease,
+		[]wapi.ValueType{
+			wapi.ValueTypeI32, // DataSource
+			wapi.ValueTypeI32, // Packet
+		},
+		[]wapi.ValueType{wapi.ValueTypeI32}, // Error
+	)
+
+	exportFunction(env, "dataSourceRelease", i.dataSourceRelease,
+		[]wapi.ValueType{
+			wapi.ValueTypeI32, // DataSource
+			wapi.ValueTypeI32, // Packet
+		},
+		[]wapi.ValueType{wapi.ValueTypeI32}, // Error
+	)
+
+	exportFunction(env, "dataArrayNew", i.dataArrayNew,
+		[]wapi.ValueType{
+			wapi.ValueTypeI32, // DataArray
+		},
+		[]wapi.ValueType{wapi.ValueTypeI32}, // Data,
+	)
+
+	exportFunction(env, "dataArrayAppend", i.dataArrayAppend,
+		[]wapi.ValueType{
+			wapi.ValueTypeI32, // DataArray
+			wapi.ValueTypeI32, // Data
+		},
+		[]wapi.ValueType{wapi.ValueTypeI32}, // Error
+	)
+
+	exportFunction(env, "dataArrayRelease", i.dataArrayRelease,
+		[]wapi.ValueType{
+			wapi.ValueTypeI32, // DataArray
+			wapi.ValueTypeI32, // Data
+		},
+		[]wapi.ValueType{wapi.ValueTypeI32}, // Error
+	)
+
+	exportFunction(env, "dataArrayLen", i.dataArrayLen,
+		[]wapi.ValueType{
+			wapi.ValueTypeI32, // DataArray
+		},
+		[]wapi.ValueType{wapi.ValueTypeI32}, // Len
+	)
+
+	exportFunction(env, "dataArrayGet", i.dataArrayGet,
+		[]wapi.ValueType{
+			wapi.ValueTypeI32, // DataArray
+			wapi.ValueTypeI32, // Index
+		},
+		[]wapi.ValueType{wapi.ValueTypeI32}, // Data,
+	)
+}
+
+// newDataSource creates a new datasource.
+// Params:
+// - stack[0] is the name of the datasource (string encoded)
+// - stack[1] is the type of the datasource
+// Return value:
+// - DataSource handle on success, 0 on error
+func (i *wasmOperatorInstance) newDataSource(ctx context.Context, m wapi.Module, stack []uint64) {
+	dsNamePtr := stack[0]
+	dsType := wapi.DecodeU32(stack[1])
+	// TODO: add type max?
+	if dsType > uint32(datasource.TypeMetrics) {
+		i.logger.Warnf("newDataSource: invalid datasource type %d", dsType)
+		stack[0] = 0
+		return
+	}
```

Comment:
IMHO this should be handled by the `RegisterDataSource()` below.

Reply:
The issue is that that API doesn't return an error now:

https://github.com/inspektor-gadget/inspektor-gadget/blob/9684c5b0127191a71d0326abd9d8bf08264627a4/pkg/datasource/data.go#L165

I'll handle it after merging this PR.

Reply:
Opened #3075